macrosynergy/download/dq_api.py

```diff
@@ -0,0 +1,749 @@
+
+        :param <int> no_tickers: number of tickers requested.
+
+        :return <float> delay: internally computed value.
+        """
+
+        if not floor(no_tickers / 100):
+            delay = 0.05
+        elif not floor(no_tickers / 1000):
+            delay = 0.2
+        else:
+            delay = 0.3
+
+        return delay
+
+    def get_ts_expression(
+        self, expression, original_metrics, suppress_warning, **kwargs
+    ):
+        """
+        Main driver function. Receives the Tickers and returns the respective dataframe.
+
+        :param <List[str]> expression: categories & respective cross-sections requested.
+        :param <List[str]> original_metrics: List of required metrics: the returned
+            DataFrame will reflect the order of the received List.
+        :param <bool> suppress_warning: required for debugging.
+        :param <dict> kwargs: dictionary of additional arguments.
+
+        :return: <pd.DataFrame> df: ['cid', 'xcat', 'real_date'] + [original_metrics].
+        """
+        if self.heartbeat:
+            logger.info("Checking connection using heartbeat")
+            clause, results = self.check_connection()
+        else:
+            clause, results = True, None
+
+        if not clause:
+            logger.error(f"Connection failed. Error message: {results}.")
+            return None
+
+        c_delay = self.delay_compute(len(expression))
+        results = None
+
+        while results is None:
+            results = self._request(
+                endpoint="/expressions/time-series",
+                tickers=expression,
+                params={},
+                delay=c_delay,
+                **kwargs,
+            )
+            c_delay += 0.1
+
+        results, error_tickers, error_messages = results
+
+        unavailable_expressions, invalid_expressions = [], []
+        for i, res in enumerate(results):
+            if res["attributes"][0]["time-series"] is None:
+                if (
+                    res["attributes"][0]["message"].strip()
+                    == "FAILED - Error in parsing JSON data"
+                ):
+                    invalid_expressions.append(res["attributes"][0]["expression"])
+                elif (
+                    res["attributes"][0]["message"]
+                    .strip()
+                    .startswith("No data available for")
+                ):
+                    unavailable_expressions.append(res["attributes"][0]["expression"])
+
+        valid_results_count = (
+            len(results) - len(invalid_expressions) - len(unavailable_expressions)
+        )
+        logger.warning(f"Invalid expressions: {', '.join(invalid_expressions)}")
+        logger.warning(f"Number of invalid expressions: {len(invalid_expressions)}")
+        logger.warning(f"Unavailable expressions: {', '.join(unavailable_expressions)}")
+        logger.warning(
+            f"Number of unavailable expressions: {len(unavailable_expressions)}"
+        )
```

Comment:
@lsimonsen; Are these conditions valid checks for unavailable & invalid expressions?